### ISSUE TITLE
Subtracting a datetime from numpy datetime reports an error

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#11135]
**Describe the bug**

If you have a `numpy.datetime64` and subtract a `datetime.datetime` (or a subclass of `datetime.datetime`), `pyright` reports an error.  `mypy` does not.

**Code or Screenshots**

```python
import numpy as np
import datetime as dt
from typing import reveal_type


class Foo(dt.datetime):
    pass


ts_np = np.datetime64("2021-01-01 03:24")

foo = Foo(2021, 3, 1)

subfoo = ts_np - foo
reveal_type(subfoo)
print(subfoo, " type ", type(subfoo))

ts_dt = dt.datetime(2020, 12, 1, 3, 10)
subdt = ts_np - ts_dt
reveal_type(subdt)
print(subdt, " type ", type(subdt))
```

numpy version 2.3.5

**VS Code extension or command-line**

pyright 1.1.407 using defaults, and on command line.

pyright output:
```text
c:\Code\Misc\pytypes\npsubdt.py
  c:\Code\Misc\pytypes\npsubdt.py:14:10 - error: Operator "-" not supported for types "datetime64[date | int | None]" and "Foo" (reportOperatorIssue)
  c:\Code\Misc\pytypes\npsubdt.py:15:13 - information: Type of "subfoo" is "Unknown"
  c:\Code\Misc\pytypes\npsubdt.py:19:9 - error: Operator "-" not supported for types "datetime64[date | int | None]" and "datetime" (reportOperatorIssue)
  c:\Code\Misc\pytypes\npsubdt.py:20:13 - information: Type of "subdt" is "Unknown"
2 errors, 0 warnings, 2 informations
```

mypy output:
```text
npsubdt.py:15: note: Revealed type is "datetime.timedelta"
npsubdt.py:20: note: Revealed type is "datetime.timedelta"
Success: no issues found in 1 source file
```

**Comment by jorenham** [comment]
Here's the relevant overload: 

```pyi
    @overload
    def __sub__(self: datetime64[_AnyDate], x: _AnyDate, /) -> dt.timedelta: ...
```

<https://github.com/numpy/numpy/blob/250285f3cefd950e3a41e049a8733a3f6a9b920c/numpy/__init__.pyi#L5625-L5626>

where `_AnyDate` is defined as

```pyi
_AnyDate = TypeVar("_AnyDate", dt.date, dt.datetime)
```

<https://github.com/numpy/numpy/blob/250285f3cefd950e3a41e049a8733a3f6a9b920c/numpy/__init__.pyi#L749>

Which for `datetime64[date | int | None]` (i.e. what `np.datetime64` defaults to) indeed doesn't match, since `date | int | None` is not assignable to either `date` or `datetime`.

So this one is on me; the default of the `datetime64` type parameter should be `Any` (for the lack of an `AnyOf`) rather than `date | int | None`. 
I tried changing this a while back, but ran into #10232 (and a similar issue with mypy), which led to more problems than it solved. But I think I'll take another looks and see if I'm able to work around it this time (e.g. by using the `Never` trick).

**Comment by jorenham** [comment]
I have opened https://github.com/numpy/numpy/pull/30364 to address this.
